internal/pkg/agent/application/coordinator/state.go

```diff
@@ -0,0 +1,331 @@
+		cs.changed()
+	}
+}
+
+// State returns the current state for the coordinator.
+func (cs *coordinatorState) State() (s State) {
+	cs.mx.RLock()
+	s.State = cs.state
+	s.Message = cs.message
+	s.FleetState = cs.fleetState
+	s.FleetMessage = cs.fleetMessage
+	s.LogLevel = cs.logLevel
+	overrideState := cs.overrideState
+	cs.mx.RUnlock()
+
+	// copy component states for PIT
+	cs.compStatesMx.RLock()
+	compStates := make([]runtime.ComponentComponentState, len(cs.compStates))
+	copy(compStates, cs.compStates)
+	cs.compStatesMx.RUnlock()
+	s.Components = compStates
+
+	if overrideState != nil {
+		// state has been overridden due to an action that is occurring
+		s.State = overrideState.state
+		s.Message = overrideState.message
+	} else if s.State == agentclient.Healthy {
+		// if any of the managers are reporting an error then something is wrong
+		// or
+		// coordinator overall is reported is healthy; in the case any component or unit is not healthy then we report
+		// as degraded because we are not fully healthy
+		cs.mgrMx.RLock()
+		defer cs.mgrMx.RUnlock()
+		if cs.runtimeMgrErr != nil {
+			s.State = agentclient.Failed
+			s.Message = cs.runtimeMgrErr.Error()
+		} else if cs.configMgrErr != nil {
+			s.State = agentclient.Failed
+			s.Message = cs.configMgrErr.Error()
+		} else if cs.actionsErr != nil {
+			s.State = agentclient.Failed
+			s.Message = cs.actionsErr.Error()
+		} else if cs.varsMgrErr != nil {
+			s.State = agentclient.Failed
+			s.Message = cs.varsMgrErr.Error()
+		} else if hasState(s.Components, client.UnitStateFailed) {
+			s.State = agentclient.Degraded
+			s.Message = "1 or more components/units in a failed state"
+		} else if hasState(s.Components, client.UnitStateDegraded) {
+			s.State = agentclient.Degraded
+			s.Message = "1 or more components/units in a degraded state"
+		}
+	}
+	return s
+}
+
+// Subscribe subscribes to changes in the coordinator state.
+//
+// This provides the current state at the time of first subscription. Cancelling the context
+// results in the subscription being unsubscribed.
+//
+// Note: Not reading from a subscription channel will cause the Coordinator to block.
+func (cs *coordinatorState) Subscribe(ctx context.Context) *StateSubscription {
+	sub := newStateSubscription(ctx, cs)
+
+	// send initial state
+	state := cs.State()
+	go func() {
+		select {
+		case <-ctx.Done():
+			return
+		case sub.ch <- state:
+		}
+	}()
+
+	// add subscription for future changes
+	cs.subMx.Lock()
+	cs.subscribe = append(cs.subscribe, sub)
+	cs.subMx.Unlock()
+
+	go func() {
+		<-ctx.Done()
+
+		// unsubscribe
+		cs.subMx.Lock()
+		defer cs.subMx.Unlock()
+		for i, s := range cs.subscribe {
+			if sub == s {
+				cs.subscribe = append(cs.subscribe[:i], cs.subscribe[i+1:]...)
+				return
+			}
+		}
+	}()
+
+	return sub
+}
+
+func (cs *coordinatorState) changed() {
+	cs.sendState(cs.State())
+}
+
+func (cs *coordinatorState) sendState(state State) {
+	cs.subMx.RLock()
+	defer cs.subMx.RUnlock()
```

Comment:
Holding this lock while doing a channel write seems a bit dangerous, what if the write blocks?

Reply:
Ah, this line is one of the reasons for that comment. 

Given that this only really matters when the agent is in testing mode, I would be biased towards copying the `subscribe` array. I suspect the majority of tests will only have 1 subscriber, and I'd rather allocate unnecessarily in test mode than potentially have time wasted debugging a deadlock.

Reply:
We still need to hold a lock when sending across the channel I believe to ensure that the order is always ensured. This is also not just used during testing. We should use the `StateWatch` GRPC for the `elastic-agent status` command when we add a `--watch`, that provides a nice way of watching

Reply:
Yeah we can block on the `sub.ch <- state:` write regardless of if we copy the slice of subscriptions first.

At some point we have to attempt to write to this channel and `<-sub.ctx.Done()` should be a reasonable guard against this.

The only more paranoid thing we could do to protect the coordinator is to add a timeout when attempting to write to the subscription channels. This would lose a state for the subscription but wouldn't block the coordinator.

I think that might be worth implementing, because if the coordinator gets blocked the agent doesn't work at all.

Reply:
I added a timer of 1 second to ensure that if it did get blocked it would not block forever.

Reply:
Thanks, I am a bit more paranoid than usual about potential deadlocks now that we have two PRs up addressing deadlocks in other places.